src/util/psbt/mod.rs

```diff
@@ -210,6 +213,234 @@ impl PartiallySignedTransaction {
 
         Ok(())
     }
+
+    /// Signs the input at `input_index` with private key `sk`.
+    ///
+    /// # Errors
+    /// If an attempt is made to sign a taproot input.
+    pub fn sign<C: Signing>(&mut self, sk: &PrivateKey, input_index: usize, secp: &Secp256k1<C>) -> Result<(), SighashError> {
+        self.check_index_is_within_bounds(input_index)?;
+
+        match self.signing_algorithm(input_index)? {
+            Algo::Ecdsa => self.sign_ecdsa(sk, input_index, secp),
+            Algo::Schnorr => Err(SighashError::TaprootNotYetSupported),
+        }
+    }
+
+    fn sign_ecdsa<C: Signing>(&mut self, sk: &PrivateKey, input_index: usize, secp: &Secp256k1<C>) -> Result<(), SighashError> {
```

Comment:
Psbts have enough information to determine what type of output is being spent and what type of signature algorithm it needs to use. 

In rust-miniscript, we have a unified method for getting a signature hash. https://docs.rs/miniscript/7.0.0/miniscript/psbt/trait.PsbtExt.html#tymethod.sighash_msg

I only made the API for sighash and not for signing because ecdsa one requires a `secp::PrivateKey` and schnorr one requires a `KeyPair`. But that I think about it, both of them are representing the same type, and it should not be an issue to keep a single API.

However, in practice, everyone uses BIP32 keys and that unifies both of the cases. 

How about we
- Make a unified API for getting a getting signature hash same as one in rust-miniscript. 
- Have a sign API that works with `PrivateKey`, converting to KeyPair for Schnorr signature hash. This is useful for non-32 BIP wallets and for signing with custom-derived keys.
- Make a single API that take inputs from a BIP32 private key, and looks up the `OutputType` and `signature_algo`. Derives the corresponding key from `PSBT_BIP32_*` fields and signs for it. 
	- For taproot, this would involve first figuring out if the key is an internal key. If it is an internal key, we should use the `key_spend` algorithm.
	- Otherwise, we want the user to supply which leaf they want to sign as an additional `Option<TapLeafHash>` parameter. In rust-miniscript, we ignore this parameter if OutputType is not Taproot. If None, use keyhash, or sign a script spend. 
	
Miniscript descriptors are able to give more detailed output here because they can also check and understand whether the leaf hash is correct and if the key is actually present or not. But this is not required here, and is not possible to do in a general case.

Reply:
How about:

```rust
// feel free to suggest a better name
pub trait PrivateKeyConvert {
    fn to_key_pair(self, context: &Secp256k1) -> KeyPair;
    fn to_private_key(self) -> PrivateKey;
}

impl PrivateKeyConvert for KeyPair {
    fn  into_key_pair(self, context: &Secp256k1) -> KeyPair {
        self
    }

    fn to_private_key(self) -> PrivateKey {
        self.into()
    }
}

impl PrivateKeyConvert for PrivateKey {
    fn  into_key_pair(self, context: &Secp256k1) -> KeyPair {
        KeyPair::from_secret_key(context, self)
    }

    fn to_private_key(self) -> PrivateKey {
        self
    }
}

impl Psbt {
    fn sign<K: ConvertPrivateKey>(&self, secp: Secp256k1, key: K /*, rest of args */) {
        // ...
        if input_is_ecdsa {
            let key = key.to_private_key();
            // ...
        } else {
            let key = key.to_key_pair(secp);
        }
    }
}
```

This saves operations if the user already has `KeyPair` while having single method.

However, this naive impl would cause monomorphisation of a large function so perhaps prefer this:

```rust
// perhaps name it differently and add #[non_exhaustive] in case new algos come?
pub enum PrivateKeyOrPair {
    JustKey(PrivateKey),
    Pair(KeyPair),
}

impl From<PrivateKey> for PrivateKeyOrPair { /* hopefully obvious */ }
impl From<KeyPair> for PrivateKeyOrPair { /* hopefully obvious */ }

impl Psbt {
    fn sign_non_generic(&self, secp: Secp256k1, key: PrivateKeyOrPair /*, rest of args */) {
        // ...
    }

    pub fn sign<K: Into<PrivateKeyOrPair>>(&self, secp: Secp256k1, key: K /*, rest of args */) {
        self.sign_non_generic(secp, key.into() /*, rest of args */)
    }
}
```

Adds a branch for each input but reduces the amount of code. Since signing is costly anyway, this is probably preferred.

Reply:
Awesome, thanks man. I'll have a play with it.